internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -242,6 +256,61 @@ func (r *BMACReconciler) reconcileAgentSpec(bmh *bmh_v1alpha1.BareMetalHost, age
 	return reconcileComplete{true}
 }
 
+// This function determines if the detached annotation should be added to the BMH
+// if the BMH is a SNO.
+//
+// The detached annotation is added if the installation of the agent associated with
+// the host has either started or has completed.
+func (r *BMACReconciler) updateBMHDetachedAnnotationIfSNO(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent) reconcileResult {
+
+	bmhAnnotations := bmh.ObjectMeta.GetAnnotations()
+	// Annotation already set to true
+	if bmhAnnotations[BMH_DETACHED_ANNOTATION] == "true" {
+		return reconcileComplete{}
+	}
+
+	// No updates if agent does not have a ClusterDeployment name
+	if agent.Spec.ClusterDeploymentName == nil {
+		return reconcileComplete{}
+	}
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		return reconcileError{err}
+	}
+
+	// If ClusterDeployment is not SNO do nothing
+	// A ClusterDeployment is SNO when ControlPlaneAgents == 1 and WorkerAgents == 0, see isSNO in
+	// https://github.com/openshift/assisted-service/blob/master/internal/controller/controllers/clusterdeployments_controller.go
+	if !(clusterDeployment.Spec.Provisioning.InstallStrategy.Agent.ProvisionRequirements.ControlPlaneAgents == 1 &&
+		clusterDeployment.Spec.Provisioning.InstallStrategy.Agent.ProvisionRequirements.WorkerAgents == 0) {
+		return reconcileComplete{}
+	}
```

Comment:
I think we may want this for all types of setups (SNO, spoke clusters) that are managed by BMAC. 

@rwsu could you elaborate on why you think this check is needed? Maybe I'm missing something

Reply:
The discussion on slack talked about the SNO scenario and I incorrectly limited the scope of this to SNO. I missed your comment about applying the annotation regardless of cluster type. These ClusterDeployment checks have been removed.